### ISSUE TITLE
14.3.37 - chef_gem fails due to dependent libraries not found 

### ISSUE THREAD

**Issue by hgregorian** [chef/chef#7511]
## Description
Libraries from installed packages not being found by gem install which depends on them.

## Chef Version
14.3.37 (narrowed down to a change between 14.3.17, which is working, and 14.3.19, which is not)

## Platform Version
CentOS 7.5 (duplicated on RHEL 6.9 and RHEL 7.5 as well)

## Replication Case
- choose gem which requires library to be installed as a prerequisite (ruby-augeas)
- ensure that devel package is installed prior to gem install (augeas-devel)
- observe failure
- multiple chef-client runs do not fix the issue

## Client Output

```
       Starting Chef Client, version 14.3.37
       Creating a new client identity for default-centos-75 using the validator key.
       resolving cookbooks for run list: ["dwa_augeas::default"]
       Synchronizing Cookbooks:
         - dwa_augeas (0.3.1)
         - build-essential (8.1.1)
         - seven_zip (3.0.0)
         - mingw (2.1.0)
         - windows (5.0.0)
         - augeas (0.5.4)
       Installing Cookbook Gems:
       Compiling Cookbooks...
       Converging 3 resources
       Recipe: dwa_augeas::default
         * yum_package[ruby-augeas dependencies] action install
           - install version 0:1.4.0-5.el7_5.1.x86_64 of package augeas-devel
           - install version 0:4.8.5-28.el7_5.1.x86_64 of package gcc
         * chef_gem[ruby-augeas] action install
           - Would install version 0.5.0 of package ruby-augeas
         * yum_package[ruby-augeas dependencies] action install (up to date)
         * chef_gem[ruby-augeas] action install
           
           ================================================================================
           Error executing action `install` on resource 'chef_gem[ruby-augeas]'
           ================================================================================
           
           Mixlib::ShellOut::ShellCommandFailed
           ------------------------------------
           Expected process to exit with [0], but received '1'
           ---- Begin output of /opt/chef/embedded/bin/gem install ruby-augeas -q --no-rdoc --no-ri -v "0.5.0" --source=https://www.rubygems.org ----
           STDOUT: Building native extensions. This could take a while...
           STDERR: ERROR:  Error installing ruby-augeas:
           	ERROR: Failed to build gem native extension.
           
        current directory: /opt/chef/embedded/lib/ruby/gems/2.5.0/gems/ruby-augeas-0.5.0/ext/augeas
           /opt/chef/embedded/bin/ruby -r ./siteconf20180727-3428-pb1tph.rb extconf.rb
           *** extconf.rb failed ***
           Could not create Makefile due to some reason, probably lack of necessary
           libraries and/or headers.  Check the mkmf.log file for more details.  You may
           need configuration options.
           
           Provided configuration options:
           	--with-opt-dir
           	--with-opt-include
           	--without-opt-include=${opt-dir}/include
           	--with-opt-lib
           	--without-opt-lib=${opt-dir}/lib
           	--with-make-prog
           	--without-make-prog
           	--srcdir=.
           	--curdir
           	--ruby=/opt/chef/embedded/bin/$(RUBY_BASE_NAME)
           	--with-augeas-config
           	--without-augeas-config
           	--with-pkg-config
           	--without-pkg-config
           extconf.rb:27:in `<main>': augeas-devel not installed (RuntimeError)
           
           To see why this extension failed to compile, please check the mkmf.log which can be found here:
           
             /opt/chef/embedded/lib/ruby/gems/2.5.0/extensions/x86_64-linux/2.5.0/ruby-augeas-0.5.0/mkmf.log
           
           extconf failed, exit code 1
           
           Gem files will remain installed in /opt/chef/embedded/lib/ruby/gems/2.5.0/gems/ruby-augeas-0.5.0 for inspection.
           Results logged to /opt/chef/embedded/lib/ruby/gems/2.5.0/extensions/x86_64-linux/2.5.0/ruby-augeas-0.5.0/gem_make.out
           ---- End output of /opt/chef/embedded/bin/gem install ruby-augeas -q --no-rdoc --no-ri -v "0.5.0" --source=https://www.rubygems.org ----
           Ran /opt/chef/embedded/bin/gem install ruby-augeas -q --no-rdoc --no-ri -v "0.5.0" --source=https://www.rubygems.org returned 1
           
           Resource Declaration:
           ---------------------
           # In /tmp/kitchen/cache/cookbooks/dwa_augeas/recipes/default.rb
           
            11: chef_gem 'ruby-augeas' do
            12:   compile_time false if respond_to?(:compile_time)
            13:   notifies :install, 'package[ruby-augeas dependencies]', :before
            14: end
            15: 
           
           Compiled Resource:
           ------------------
           # Declared in /tmp/kitchen/cache/cookbooks/dwa_augeas/recipes/default.rb:11:in `from_file'
           
           chef_gem("ruby-augeas") do
             package_name "ruby-augeas"
             action [:install]
             updated true
             default_guard_interpreter :default
             declared_type :chef_gem
             cookbook_name "dwa_augeas"
             recipe_name "default"
           end
           
           System Info:
           ------------
           chef_version=14.3.37
           platform=centos
           platform_version=7.5.1804
           ruby=ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux]
           program_name=/opt/chef/bin/chef-client
           executable=/opt/chef/bin/chef-client
           
       
       Running handlers:
       [2018-07-27T20:22:18+00:00] ERROR: Running exception handlers
       [2018-07-27T20:22:18+00:00] ERROR: Running exception handlers
       Running handlers complete
       [2018-07-27T20:22:18+00:00] ERROR: Exception handlers complete
       [2018-07-27T20:22:18+00:00] ERROR: Exception handlers complete
       Chef Client failed. 2 resources updated in 32 seconds
       [2018-07-27T20:22:18+00:00] FATAL: Stacktrace dumped to /tmp/kitchen/cache/chef-stacktrace.out
       [2018-07-27T20:22:18+00:00] FATAL: Stacktrace dumped to /tmp/kitchen/cache/chef-stacktrace.out
       [2018-07-27T20:22:18+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
       [2018-07-27T20:22:18+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
       [2018-07-27T20:22:18+00:00] FATAL: Mixlib::ShellOut::ShellCommandFailed: chef_gem[ruby-augeas] (dwa_augeas::default line 11) had an error: Mixlib::ShellOut::ShellCommandFailed: Expected process to exit with [0], but received '1'
       ---- Begin output of /opt/chef/embedded/bin/gem install ruby-augeas -q --no-rdoc --no-ri -v "0.5.0" --source=https://www.rubygems.org ----
       STDOUT: Building native extensions. This could take a while...
       STDERR: ERROR:  Error installing ruby-augeas:
       	ERROR: Failed to build gem native extension.
       
           current directory: /opt/chef/embedded/lib/ruby/gems/2.5.0/gems/ruby-augeas-0.5.0/ext/augeas
       /opt/chef/embedded/bin/ruby -r ./siteconf20180727-3428-pb1tph.rb extconf.rb
       *** extconf.rb failed ***
       Could not create Makefile due to some reason, probably lack of necessary
       libraries and/or headers.  Check the mkmf.log file for more details.  You may
       need configuration options.
       
       Provided configuration options:
       	--with-opt-dir
       	--with-opt-include
       	--without-opt-include=${opt-dir}/include
       	--with-opt-lib
       	--without-opt-lib=${opt-dir}/lib
       	--with-make-prog
       	--without-make-prog
       	--srcdir=.
       	--curdir
       	--ruby=/opt/chef/embedded/bin/$(RUBY_BASE_NAME)
       	--with-augeas-config
       	--without-augeas-config
       	--with-pkg-config
       	--without-pkg-config
       extconf.rb:27:in `<main>': augeas-devel not installed (RuntimeError)
       
       To see why this extension failed to compile, please check the mkmf.log which can be found here:
       
         /opt/chef/embedded/lib/ruby/gems/2.5.0/extensions/x86_64-linux/2.5.0/ruby-augeas-0.5.0/mkmf.log
       
       extconf failed, exit code 1
       
       Gem files will remain installed in /opt/chef/embedded/lib/ruby/gems/2.5.0/gems/ruby-augeas-0.5.0 for inspection.
       Results logged to /opt/chef/embedded/lib/ruby/gems/2.5.0/extensions/x86_64-linux/2.5.0/ruby-augeas-0.5.0/gem_make.out
       ---- End output of /opt/chef/embedded/bin/gem install ruby-augeas -q --no-rdoc --no-ri -v "0.5.0" --source=https://www.rubygems.org ----
       Ran /opt/chef/embedded/bin/gem install ruby-augeas -q --no-rdoc --no-ri -v "0.5.0" --source=https://www.rubygems.org returned 1
       [2018-07-27T20:22:18+00:00] FATAL: Mixlib::ShellOut::ShellCommandFailed: chef_gem[ruby-augeas] (dwa_augeas::default line 11) had an error: Mixlib::ShellOut::ShellCommandFailed: Expected process to exit with [0], but received '1'
       ---- Begin output of /opt/chef/embedded/bin/gem install ruby-augeas -q --no-rdoc --no-ri -v "0.5.0" --source=https://www.rubygems.org ----
       STDOUT: Building native extensions. This could take a while...
       STDERR: ERROR:  Error installing ruby-augeas:
       	ERROR: Failed to build gem native extension.
       
           current directory: /opt/chef/embedded/lib/ruby/gems/2.5.0/gems/ruby-augeas-0.5.0/ext/augeas
       /opt/chef/embedded/bin/ruby -r ./siteconf20180727-3428-pb1tph.rb extconf.rb
       *** extconf.rb failed ***
       Could not create Makefile due to some reason, probably lack of necessary
       libraries and/or headers.  Check the mkmf.log file for more details.  You may
       need configuration options.
       
       Provided configuration options:
       	--with-opt-dir
       	--with-opt-include
       	--without-opt-include=${opt-dir}/include
       	--with-opt-lib
       	--without-opt-lib=${opt-dir}/lib
       	--with-make-prog
       	--without-make-prog
       	--srcdir=.
       	--curdir
       	--ruby=/opt/chef/embedded/bin/$(RUBY_BASE_NAME)
       	--with-augeas-config
       	--without-augeas-config
       	--with-pkg-config
       	--without-pkg-config
       extconf.rb:27:in `<main>': augeas-devel not installed (RuntimeError)
       
       To see why this extension failed to compile, please check the mkmf.log which can be found here:
       
         /opt/chef/embedded/lib/ruby/gems/2.5.0/extensions/x86_64-linux/2.5.0/ruby-augeas-0.5.0/mkmf.log
       
       extconf failed, exit code 1
       
       Gem files will remain installed in /opt/chef/embedded/lib/ruby/gems/2.5.0/gems/ruby-augeas-0.5.0 for inspection.
       Results logged to /opt/chef/embedded/lib/ruby/gems/2.5.0/extensions/x86_64-linux/2.5.0/ruby-augeas-0.5.0/gem_make.out
       ---- End output of /opt/chef/embedded/bin/gem install ruby-augeas -q --no-rdoc --no-ri -v "0.5.0" --source=https://www.rubygems.org ----
       Ran /opt/chef/embedded/bin/gem install ruby-augeas -q --no-rdoc --no-ri -v "0.5.0" --source=https://www.rubygems.org returned 1
```

## Stacktrace

```
>>>> Caused by Mixlib::ShellOut::ShellCommandFailed: Expected process to exit with [0], but received '1'
---- Begin output of /opt/chef/embedded/bin/gem install ruby-augeas -q --no-rdoc --no-ri -v "0.5.0" --source=https://www.rubygems.org ----
STDOUT: Building native extensions. This could take a while...
STDERR: ERROR:  Error installing ruby-augeas:
	ERROR: Failed to build gem native extension.

    current directory: /opt/chef/embedded/lib/ruby/gems/2.5.0/gems/ruby-augeas-0.5.0/ext/augeas
/opt/chef/embedded/bin/ruby -r ./siteconf20180727-3428-pb1tph.rb extconf.rb
*** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
	--with-opt-dir
	--with-opt-include
	--without-opt-include=${opt-dir}/include
	--with-opt-lib
	--without-opt-lib=${opt-dir}/lib
	--with-make-prog
	--without-make-prog
	--srcdir=.
	--curdir
	--ruby=/opt/chef/embedded/bin/$(RUBY_BASE_NAME)
	--with-augeas-config
	--without-augeas-config
	--with-pkg-config
	--without-pkg-config
extconf.rb:27:in `<main>': augeas-devel not installed (RuntimeError)

To see why this extension failed to compile, please check the mkmf.log which can be found here:

  /opt/chef/embedded/lib/ruby/gems/2.5.0/extensions/x86_64-linux/2.5.0/ruby-augeas-0.5.0/mkmf.log

extconf failed, exit code 1

Gem files will remain installed in /opt/chef/embedded/lib/ruby/gems/2.5.0/gems/ruby-augeas-0.5.0 for inspection.
Results logged to /opt/chef/embedded/lib/ruby/gems/2.5.0/extensions/x86_64-linux/2.5.0/ruby-augeas-0.5.0/gem_make.out
---- End output of /opt/chef/embedded/bin/gem install ruby-augeas -q --no-rdoc --no-ri -v "0.5.0" --source=https://www.rubygems.org ----
Ran /opt/chef/embedded/bin/gem install ruby-augeas -q --no-rdoc --no-ri -v "0.5.0" --source=https://www.rubygems.org returned 1
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/mixlib-shellout-2.3.2/lib/mixlib/shellout.rb:293:in `invalid!'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/mixlib-shellout-2.3.2/lib/mixlib/shellout.rb:280:in `error!'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/mixin/shell_out.rb:197:in `shell_out_compacted!'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/mixin/shell_out.rb:124:in `shell_out!'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/provider/package/rubygems.rb:550:in `install_via_gem_command'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/provider/package/rubygems.rb:531:in `install_package'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/provider/package.rb:102:in `block (3 levels) in <class:Package>'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/provider/package.rb:271:in `multipackage_api_adapter'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/provider/package.rb:101:in `block (2 levels) in <class:Package>'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/mixin/why_run.rb:51:in `add_action'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/provider.rb:227:in `converge_by'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/provider/package.rb:100:in `block in <class:Package>'
(eval):2:in `block in action_install'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/provider.rb:236:in `instance_eval'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/provider.rb:236:in `compile_and_converge_action'
(eval):2:in `action_install'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/provider.rb:182:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/resource.rb:577:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/runner.rb:70:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/runner.rb:98:in `block (2 levels) in converge'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/runner.rb:98:in `each'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/runner.rb:98:in `block in converge'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/resource_collection/resource_list.rb:94:in `block in execute_each_resource'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/resource_collection/stepable_iterator.rb:114:in `call_iterator_block'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/resource_collection/stepable_iterator.rb:103:in `iterate'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/resource_collection/resource_list.rb:92:in `execute_each_resource'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/runner.rb:97:in `converge'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/client.rb:720:in `block in converge'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/client.rb:715:in `catch'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/client.rb:715:in `converge'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/client.rb:754:in `converge_and_save'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/client.rb:286:in `run'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/application.rb:303:in `run_with_graceful_exit_option'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/application.rb:279:in `block in run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/local_mode.rb:44:in `with_server_connectivity'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/application.rb:261:in `run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/application/client.rb:440:in `run_application'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/lib/chef/application.rb:66:in `run'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.3.37/bin/chef-client:25:in `<top (required)>'
/opt/chef/bin/chef-client:75:in `load'
/opt/chef/bin/chef-client:75:in `<main>'
```

**Comment by knightorc** [comment]
You're not alone.  I'm currently working on migrating some cookbooks that depend on `rvm` (cookbook that needs too much work for Chef 13 or 14 support) to chef_rvm (mostly supports 13 and greater + some tweaks which I hope to give back to the community).

We've got one passenger app that requires ruby-augeas and runs ruby via rvm not system ruby. 

chef 12.x.x  has no issues
chef 13.6 - 13.10.0 I've tried everything I can possibly thing of to (path, env variables, etc) to expose the installed augeas-devel libraries with no success.

If I magically figure out what is needed/the magic trick I'll post again.

Sadly this is the last challenge that is blocking our linux migration from 12.22.3 -> 13.10.0
Our windows nodes have been on 13.10.0 for 9 months.

**Comment by nhuff** [comment]
This has to do with the PATH that is getting set when gem is called.  In newer version of chef /opt/chef/embedded/bin is prepended to the PATH environment which causes it to use the embedded pkg-config instead of the system package config.  The embedded pkg-config only searches the embedded library directories and so can't find the augeas libraries installed on the system. Not sure what the best solution is here though.

**Comment by lamont-granquist** [comment]
Did this issue get hijacked?  because the original submitter quite clearly mentioned that this was a change in behavior in 14.3.18 or 14.3.19, not in Chef > 12.

The Chef > 12 is certainly easier to understand where it comes from though.

**Comment by lamont-granquist** [comment]
The original issue here seems to have been caused by #7372, presumably from the change from `shell_out_with_timeout!` to `shell_out!` in the rubygems provider, but I can't see how that changed PATH behavior.  That should have been fully backwards compatible since `shell_out_with_timeout!` used to call `shell_out!` and the base behavior of `shell_out!` didn't change.

The fact that the rubygems provider explicitly passes `env: nil` is weird.  I don't know why that is there or if there might be some odd edge case in its behavior.

**Comment by nhuff** [comment]
Change in behavior happened after 14.2.0.  14.3.37, next version available on download page has the issue.

**Comment by nhuff** [comment]
Still appears to be an issue in 15.5.17.

**Comment by lamont-granquist** [comment]
15.5.17 shouldn't have had libtool or pkg-config in it, I wonder how it crept back in

**Comment by lamont-granquist** [comment]
it didn't ship is the answer.  15.5.23 from current may have fixed this via removing libtool+pkg-config (once it builds).

**Comment by nhuff** [comment]
Seems to be fixed in 15.6.10

**Comment by lamont-granquist** [comment]
\o/

finally got rid of all the problems that libtool+pkg-config cause.
